# Profiling — Czech Bank (Berka)

In [32]:
import pandas as pd

pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 200)

DATA = '../data/bronze/'
account  = pd.read_csv(DATA + 'account.csv', sep=';')
card     = pd.read_csv(DATA + 'card.csv', sep=';')
client   = pd.read_csv(DATA + 'client.csv', sep=';')
disp     = pd.read_csv(DATA + 'disp.csv', sep=';')
district = pd.read_csv(DATA + 'district.csv',sep=';')
loan     = pd.read_csv(DATA + 'loan.csv', sep=';')
orders   = pd.read_csv(DATA + 'order.csv', sep=';')
trans    = pd.read_csv(DATA + 'trans.csv',  sep=';', low_memory=False)

bang = {
    'account': account, 'card': card, 'client': client, 'disp': disp,
    'district': district, 'loan': loan, 'orders': orders, 'trans': trans,
}

print('Da nap 8 bang')

Da nap 8 bang


#Kiểm tra cấu trúc tất cả các bảng



In [33]:
def quet_cau_truc(ten, df):
    print('=' * 90)
    print(f'{ten.upper()}   —   {df.shape[0]:,} dong  x  {df.shape[1]} cot')
    print('=' * 90)
    display(df.head(3))
    tom_tat = pd.DataFrame({
        'kieu': df.dtypes.astype(str),
        'so_null': df.isnull().sum(),
        'so_gia_tri_khac_nhau': df.nunique(),
    })
    display(tom_tat)
    print()

In [34]:
for ten, df in bang.items():
    quet_cau_truc(ten, df)

ACCOUNT   —   4,500 dong  x  4 cot


,account_id,district_id,frequency,date
0,576,55,POPLATEK MESICNE,930101
1,3818,74,POPLATEK MESICNE,930101
2,704,55,POPLATEK MESICNE,930101


,kieu,so_null,so_gia_tri_khac_nhau
account_id,int64,0,4500
district_id,int64,0,77
frequency,object,0,3
date,int64,0,1535



CARD   —   892 dong  x  4 cot


,card_id,disp_id,type,issued
0,1005,9285,classic,931107 00:00:00
1,104,588,classic,940119 00:00:00
2,747,4915,classic,940205 00:00:00


,kieu,so_null,so_gia_tri_khac_nhau
card_id,int64,0,892
disp_id,int64,0,892
type,object,0,3
issued,object,0,607



CLIENT   —   5,369 dong  x  3 cot


,client_id,birth_number,district_id
0,1,706213,18
1,2,450204,1
2,3,406009,1


,kieu,so_null,so_gia_tri_khac_nhau
client_id,int64,0,5369
birth_number,int64,0,5019
district_id,int64,0,77



DISP   —   5,369 dong  x  4 cot


,disp_id,client_id,account_id,type
0,1,1,1,OWNER
1,2,2,2,OWNER
2,3,3,2,DISPONENT


,kieu,so_null,so_gia_tri_khac_nhau
disp_id,int64,0,5369
client_id,int64,0,5369
account_id,int64,0,4500
type,object,0,2



DISTRICT   —   77 dong  x  16 cot


,A1,A2,A3,A4,A5,A6,A7,A8,A9,A10,A11,A12,A13,A14,A15,A16
0,1,Hl.m. Praha,Prague,1204953,0,0,0,1,1,100.0,12541,0.29,0.43,167,85677,99107
1,2,Benesov,central Bohemia,88884,80,26,6,2,5,46.7,8507,1.67,1.85,132,2159,2674
2,3,Beroun,central Bohemia,75232,55,26,4,1,5,41.7,8980,1.95,2.21,111,2824,2813


,kieu,so_null,so_gia_tri_khac_nhau
A1,int64,0,77
A2,object,0,77
A3,object,0,8
A4,int64,0,77
A5,int64,0,53
A6,int64,0,36
A7,int64,0,17
A8,int64,0,6
A9,int64,0,11
A10,float64,0,70



LOAN   —   682 dong  x  7 cot


,loan_id,account_id,date,amount,duration,payments,status
0,5314,1787,930705,96396,12,8033.0,B
1,5316,1801,930711,165960,36,4610.0,A
2,6863,9188,930728,127080,60,2118.0,A


,kieu,so_null,so_gia_tri_khac_nhau
loan_id,int64,0,682
account_id,int64,0,682
date,int64,0,559
amount,int64,0,645
duration,int64,0,5
payments,float64,0,577
status,object,0,4



ORDERS   —   6,471 dong  x  6 cot


,order_id,account_id,bank_to,account_to,amount,k_symbol
0,29401,1,YZ,87144583,2452.0,SIPO
1,29402,2,ST,89597016,3372.7,UVER
2,29403,2,QR,13943797,7266.0,SIPO


,kieu,so_null,so_gia_tri_khac_nhau
order_id,int64,0,6471
account_id,int64,0,3758
bank_to,object,0,13
account_to,int64,0,6446
amount,float64,0,4412
k_symbol,object,0,5



TRANS   —   1,056,320 dong  x  10 cot


,trans_id,account_id,date,type,operation,amount,balance,k_symbol,bank,account
0,695247,2378,930101,PRIJEM,VKLAD,700.0,700.0,NaN,NaN,NaN
1,171812,576,930101,PRIJEM,VKLAD,900.0,900.0,NaN,NaN,NaN
2,207264,704,930101,PRIJEM,VKLAD,1000.0,1000.0,NaN,NaN,NaN


,kieu,so_null,so_gia_tri_khac_nhau
trans_id,int64,0,1056320
account_id,int64,0,4500
date,int64,0,2191
type,object,0,3
operation,object,183114,5
amount,float64,0,40400
balance,float64,0,542739
k_symbol,object,481881,8
bank,object,782812,13
account,float64,760931,7665


## Nhận xét
- Ở bảng DISTRICT: A12 VÀ A15 kiểu object, nhưng A13 và A16 kiểu số. Trong khi 4 cái này là 2 cặp chỉ số. 
- account.date, trans.date, loan.date đều có kiểu int, trong khi đó là ngày
- disp cps 5.369 dòng, mà account chỉ có 4.500 dòng, dữ liệu bị lệch?
- trans, operation và k_symbol có nhiều rỗng
- clinet: có 3 cột, không có ngày sinh, không có giới tính | Nhưng birth_number trông giống ngày 

Nên đi kiểm tra những điểm dữ liệu đó

# 1 — `district`: cột cùng loại nhưng khác kiểu

In [35]:
district[['A12', 'A13', 'A15', 'A16']].dtypes

A12     object
A13    float64
A15     object
A16      int64
dtype: object

Nhưng khi kiểm tra cột `so_null` của `district` toàn số 0,  không có ô nào rỗng.

In [36]:
# to_numeric voi errors='coerce' bien gia tri khong phai so thanh NaN.
# Dem xem co bao nhieu gia tri nhu vay.
for cot in ['A12', 'A15']:
    khong_phai_so = pd.to_numeric(district[cot], errors='coerce').isnull()
    print(f'{cot}: {khong_phai_so.sum()} gia tri khong ep duoc sang so')

A12: 1 gia tri khong ep duoc sang so
A15: 1 gia tri khong ep duoc sang so


In [37]:
loi = (pd.to_numeric(district['A12'], errors='coerce').isnull()
       | pd.to_numeric(district['A15'], errors='coerce').isnull())

district[loi]

,A1,A2,A3,A4,A5,A6,A7,A8,A9,A10,A11,A12,A13,A14,A15,A16
68,69,Jesenik,north Moravia,42821,4,13,5,1,3,48.4,8173,?,7.01,124,?,1358


## Nhận xét 1
- xuất hiện "?" chứ không phải số. Đây là quy ước đánh dấu dữ liệu thiếu của các file phẳng thời chưa có chuẩn NULL.

Vì vậy `isnull()` không bắt được nó, vì với pandas thì `'?'` là một chuỗi hợp lệ chứ không phải giá trị rỗng. 

Quận này còn số liệu của năm còn lại không? Và nó ảnh hưởng bao nhiêu khách hàng?

In [38]:
quan_loi = district[loi]['A1'].iloc[0]
so_tk = (account['district_id'] == quan_loi).sum()

print('Quan bi loi, id =', quan_loi)
print('So tai khoan mo o quan nay:', so_tk)
print('Tong so tai khoan         :', len(account))
print(f'Ty le: {100 * so_tk / len(account):.2f}%')

Quan bi loi, id = 69
So tai khoan mo o quan nay: 48
Tong so tai khoan         : 4500
Ty le: 1.07%


## Kết luận 1

Quận này vẫn có số liệu của năm còn lại (A13, A16 đều có giá trị). Chỉ thiếu số liệu của một năm.

**Quyết định: để NULL.** Không điền thay thế, vì điền sẽ tạo ra một phát hiện giả mà về sau không ai kiểm chứng được.


#2 — Các cột `date` đang là số nguyên

In [39]:
for ten, df in bang.items():
    for cot in df.columns:
        if any(k in cot.lower() for k in ['date', 'issued']):
            print(f'{ten}.{cot:<12} kiểu = {df[cot].dtype}    dạng = {df[cot].iloc[0]}')

account.date         kiểu = int64    dạng = 930101
card.issued       kiểu = object    dạng = 931107 00:00:00
loan.date         kiểu = int64    dạng = 930705
trans.date         kiểu = int64    dạng = 930101


Giá trị mẫu trông như `930101`. Nếu đây là ngày dạng `YYMMDD` thì tách ra phải hợp lý: hai chữ số giữa nằm trong 1–12, hai chữ số cuối nằm trong 1–31.

Viết hàm kiểm tra giả thuyết đó

In [40]:
def kiem_tra_yymmdd(ten, chuoi_so):
    nam   = chuoi_so // 10000
    thang = (chuoi_so // 100) % 100
    ngay  = chuoi_so % 100
    print(f'{ten:<22} nam {nam.min()}-{nam.max()}   '
          f'thang {thang.min()}-{thang.max()}   '
          f'ngay {ngay.min()}-{ngay.max()}')

kiem_tra_yymmdd('account.date', account['date'])
kiem_tra_yymmdd('trans.date',   trans['date'])
kiem_tra_yymmdd('loan.date',    loan['date'])

account.date           nam 93-97   thang 1-12   ngay 1-30
trans.date             nam 93-98   thang 1-12   ngay 1-31
loan.date              nam 93-98   thang 1-12   ngay 1-31


## Nhận xét 2

Cả ba cột đều thoả: tháng nằm trong 1–12, ngày nằm trong 1–31. Giả thuyết `YYMMDD` đúng

Cách khắc phục:  đổi sang kiểu `DATE` dùng `strptime(CAST(date AS VARCHAR), '%y%m%d')::DATE`.

In [41]:
thu = pd.to_datetime(trans['date'].astype(str), format='%y%m%d')

print('Giao dich som nhat :', thu.min().date())
print('Giao dich muon nhat:', thu.max().date())
print('So nam du lieu :', thu.max().year - thu.min().year + 1)

Giao dich som nhat : 1993-01-01
Giao dich muon nhat: 1998-12-31
So nam du lieu : 6


#3 — `disp` nhiều dòng hơn `account`

In [42]:
print('disp   :', len(disp))
print('account:', len(account))
print('client :', len(client))
print()
print('So account_id bi lap trong disp:', disp['account_id'].duplicated().sum())
print('So client_id  bi lap trong disp:', disp['client_id'].duplicated().sum())

disp   : 5369
account: 4500
client : 5369

So account_id bi lap trong disp: 869
So client_id  bi lap trong disp: 0


`client_id` không bị lặp, trong khi `account_id` lại bị lặp. Điều này cho thấy một tài khoản có thể được gắn với nhiều khách hàng, chứ không phải một khách hàng có nhiều tài khoản. Vì vậy,cần kiểm tra xem trung bình mỗi tài khoản đang được gắn với bao nhiêu người.


In [43]:
disp['account_id'].value_counts().value_counts().sort_index()

count
1    3631
2     869
Name: count, dtype: int64

In [44]:
# Nhin mot truong hop cu the
dem = disp['account_id'].value_counts()
vi_du = dem[dem > 1].index[0]

disp[disp['account_id'] == vi_du]

,disp_id,client_id,account_id,type
2706,2862,2862,2369,OWNER
2707,2863,2863,2369,DISPONENT


In [45]:
#Kiểm tra xem có bao nhiêu loại trong type
disp['type'].value_counts()

type
OWNER        4500
DISPONENT     869
Name: count, dtype: int64

## Nhận xét 3

Mỗi tài khoản sẽ có một người đứng tên chính (OWNER), và một số tài khoản có thêm người được ủy quyền (DISPONENT). Việc số lượng OWNER bằng số lượng tài khoản là hợp lý, vì mỗi tài khoản chỉ có một chủ.

Vì vậy, account_id bị lặp trong disp không phải do dữ liệu bị lỗi, mà là do một tài khoản có thể gắn với hai người với hai vai trò khác nhau.

Điểm này cần lưu ý khi JOIN bảng. Nếu nối trans với disp, những tài khoản có cả OWNER và DISPONENT sẽ làm mỗi giao dịch khớp với 2 dòng. Khi đó, nếu tính số giao dịch hoặc tổng tiền mà không xử lý trước, kết quả rất dễ bị đếm trùng.

In [46]:
# trans JOIN account
kq1 = trans.merge(account, on='account_id')

print('trans:', len(trans))
print('Sau khi join account:', len(kq1))
print('Tăng:', len(kq1) - len(trans))


# trans JOIN disp
kq2 = trans.merge(disp, on='account_id')

print('\ntrans:', len(trans))
print('Sau khi join disp:', len(kq2))
print('Tăng:', len(kq2) - len(trans))

trans: 1056320
Sau khi join account: 1056320
Tăng: 0

trans: 1056320
Sau khi join disp: 1262625
Tăng: 206305


Số dòng tăng lên như dự đoán

Kiểm tra số tiền 
balance: đây là số dư *sau mỗi giao dịch*, tức số dư luỹ kế. Cộng toàn bộ cột này là vô nghĩa. Muốn biết tổng số dư danh mục thì lấy số dư của **giao dịch cuối cùng** mỗi tài khoản.

Loại chỉ số này gọi là **semi-additive** — cộng được theo tài khoản nhưng không cộng được theo thời gian. 

In [47]:
so_du_cuoi = (trans
              .sort_values(['account_id', 'date', 'trans_id'])
              .groupby('account_id')['balance']
              .last()
              .reset_index())

so_du_cuoi.head()

,account_id,balance
0,1,13466.5
1,2,42628.1
2,3,51096.1
3,4,33720.7
4,5,28088.3


In [48]:
# Chỉ lấy những dòng là chủ tài khoản
disp_owner = disp[disp['type'] == 'OWNER']
# Tổng số dư đúng
tong_dung = so_du_cuoi['balance'].sum()
print('Tổng số dư đúng:', tong_dung)
# JOIN với toàn bộ disp
merge_day_du = so_du_cuoi.merge(disp, on='account_id')

tong_sai = merge_day_du['balance'].sum()

print('Tổng sau khi JOIN với disp:', tong_sai)

# JOIN chỉ với OWNER
merge_owner = so_du_cuoi.merge(disp_owner, on='account_id')

tong_owner = merge_owner['balance'].sum()

print('Tổng sau khi chỉ lấy OWNER:', tong_owner)

Tổng số dư đúng: 197140234.2
Tổng sau khi JOIN với disp: 233994805.4
Tổng sau khi chỉ lấy OWNER: 197140234.2


## Kết luận  3

**Quyết định: lọc `disp.type = 'OWNER'`.** Sau khi lọc, `disp` còn 4.500 dòng và khớp 1-1 với `account`.

Lọc về OWNER và không lấy tin tin về người ủy quyền, vì project chỉ phân tích về chủ tài khoản

#  Quét đồng nhất: giá trị của các cột phân loại

Lại quét đều cả 8 bảng. Liệt kê toàn bộ giá trị của những cột có ít giá trị khác nhau.

In [49]:
def quet_phan_loai(ten, df, nguong=12):
    for cot in df.columns:
        n = df[cot].nunique()
        if 1 < n <= nguong:
            print(f'--- {ten}.{cot}   ({n} gia tri)')
            print(df[cot].value_counts(dropna=False).to_string())
            print()

In [50]:
for ten, df in bang.items():
    quet_phan_loai(ten, df)

--- account.frequency   (3 gia tri)
frequency
POPLATEK MESICNE      4167
POPLATEK TYDNE         240
POPLATEK PO OBRATU      93

--- card.type   (3 gia tri)
type
classic    659
junior     145
gold        88

--- disp.type   (2 gia tri)
type
OWNER        4500
DISPONENT     869

--- district.A3   (8 gia tri)
A3
south Moravia      14
central Bohemia    12
east Bohemia       11
north Moravia      11
west Bohemia       10
north Bohemia      10
south Bohemia       8
Prague              1

--- district.A8   (6 gia tri)
A8
1    34
2    23
3    12
0     4
4     3
5     1

--- district.A9   (11 gia tri)
A9
6     14
4     12
7     11
5     10
8      7
9      7
10     7
1      4
11     2
3      2
2      1

--- loan.duration   (5 gia tri)
duration
60    145
24    138
48    138
12    131
36    130

--- loan.status   (4 gia tri)
status
C    403
A    203
D     45
B     31

--- orders.k_symbol   (5 gia tri)
k_symbol
SIPO        3502
            1379
UVER         717
POJISTNE     532
LEASING      341

--

## Nhận xét
**Cac ngôn ngữ của tiếng Séc** `POPLATEK MESICNE`, `VYDAJ`, `PRIJEM`, `VYBER`, `PREVOD NA UCET`, `VKLAD`, `UROK`, `SLUZBY`, `SIPO`, `DUCHOD`, `POJISTNE`, `UVER`. Vì vậy cần dịch lại cho dễ hiểu

`trans.type có **ba** giá trị trong khi đó Tiền chỉ có vào hoặc ra. 

`trans.k_symbol` vừa có `NaN` vừa có giá trị giống như khoảng trắng. Hai kiểu rỗng khác nhau trong cùng một cột

Kiểm tra `trans.type` có ba giá trị

In [51]:
pd.crosstab(
    trans['type'],
    trans['operation'].fillna('(rong)')
)

operation,(rong),PREVOD NA UCET,PREVOD Z UCTU,VKLAD,VYBER,VYBER KARTOU
type,,,,,,
PRIJEM,183114,0,65226,156743,0,0
VYBER,0,0,0,0,16666,0
VYDAJ,0,208283,0,0,418252,8036


## Nhận xét 

Có một giá trị thứ ba trong cột type, nhưng nó chỉ xuất hiện cùng một loại operation, và loại operation đó cũng đã xuất hiện ở hai giá trị còn lại.

Điều này cho thấy giá trị thứ ba có thể không phải một loại giao dịch mới, mà chỉ là một cách ghi khác của một loại đã có. Nói đơn giản là cùng một bản chất giao dịch nhưng dữ liệu lại được ghi thành hai cách khác nhau.

Để chắc chắn, cần so sánh số tiền giao dịch giữa các nhóm xem chúng có thực sự giống nhau hay không.

In [52]:
trans.groupby('type')['amount'].describe()

,count,mean,std,min,25%,50%,75%,max
type,,,,,,,,
PRIJEM,405083.0,7967.434882,11835.612067,0.0,143.5,1600.0,12087.5,74812.0
VYBER,16666.0,12516.726149,6593.286488,2001.0,6738.5,12103.0,18081.0,24997.0
VYDAJ,634571.0,4446.653531,7375.546354,0.0,43.0,2042.0,5200.0,87400.0


Vì vậy tôi chọn gộp giá trị thứ ba vào giá trị cùng nghĩa, đưa về đúng hai chiều tiền vào / tiền ra.

 — `k_symbol` có hai kiểu rỗng

In [53]:
print('NaN:', trans['k_symbol'].isna().sum())
print('Chuỗi rỗng:', (trans['k_symbol'] == '').sum())
print('Khoảng trắng:', (trans['k_symbol'] == ' ').sum())

NaN: 481881
Chuỗi rỗng: 0
Khoảng trắng: 53433


## Nhận xét
Cột k_symbol đang có hai kiểu dữ liệu được xem là rỗng: một loại là NaN, một loại là chuỗi chỉ chứa khoảng trắng. Hai loại này thực ra đều có nghĩa là không có thông tin, nhưng dữ liệu lại đang lưu thành hai dạng khác nhau.

Nếu không xử lý, khi lọc IS NOT NULL thì những dòng chỉ chứa khoảng trắng vẫn được giữ lại.

Vì vậy, tôi quyết định đưa cả hai về cùng một dạng là NULL. 

operation` rỗng rất nhiều
Trước khi xoá hay điền, phải hỏi: **cột này có áp dụng cho những dòng đó không?**

So số dòng rỗng của `operation` với số dòng của từng giá trị trong `k_symbol`:

In [54]:
# Đếm số dòng bị thiếu trong cột operation
print('operation rỗng:', trans['operation'].isnull().sum())
print()

# Đếm số dòng của từng giá trị trong cột k_symbol
# dropna=False để tính luôn cả các dòng NaN
print('Số dòng theo từng giá trị k_symbol:')
print(trans['k_symbol'].value_counts(dropna=False).to_string())

operation rỗng: 183114

Số dòng theo từng giá trị k_symbol:
k_symbol
NaN            481881
UROK           183114
SLUZBY         155832
SIPO           118065
                53433
DUCHOD          30338
POJISTNE        18500
UVER            13580
SANKC. UROK      1577


Có một điểm khá đáng chú ý: số dòng operation bị thiếu là 183.114 bằng số dòng có k_symbol = UROK. Hai con số trùng nhau làm tôi nghi ngờ rằng có thể hai nhóm này thực chất là cùng một tập dữ liệu.

Tuy nhiên, chỉ nhìn vào số lượng thì chưa đủ để kết luận. Cần lọc riêng những dòng operation bị thiếu và kiểm tra k_symbol của chúng. Nếu toàn bộ 183.114 dòng đều có k_symbol = UROK, thì có thể xác nhận rằng các giao dịch UROK chính là những giao dịch bị thiếu operation.

In [55]:
# Tìm những dòng bị thiếu operation
thieu_operation = trans['operation'].isnull()

# Kiểm tra k_symbol của những dòng này
print('k_symbol trong các giao dịch bị thiếu operation:')
print(trans.loc[thieu_operation, 'k_symbol']
      .value_counts(dropna=False)
      .to_string())

k_symbol trong các giao dịch bị thiếu operation:
k_symbol
UROK    183114


In [56]:
# Xem đặc điểm của các giao dịch bị thiếu operation không?

print('Loại giao dịch:')
print(trans[thieu_operation]['type'].value_counts().to_string())

print()

print('Thống kê số tiền giao dịch:')
print(trans[thieu_operation]['amount'].describe().to_string())

Loại giao dịch:
type
PRIJEM    183114

Thống kê số tiền giao dịch:
count    183114.000000
mean        150.019603
std          80.510640
min           0.000000
25%          89.700000
50%         133.000000
75%         195.600000
max         565.500000


## Nhận xét
- Số dòng `operation` rỗng trùng **chính xác** với số dòng của một giá trị `k_symbol` cụ thể
- Toàn bộ những dòng đó đều là **tiền vào**
- ố tiền **nhỏ và đều**, khác hẳn giao dịch thông thường

Có thể suy ra: đây là khoản **lãi ngân hàng tự động ghi có** cho khách. Cột `operation` mô tả *hành động của khách hàng* — rút, nộp, chuyển. Khoản lãi này do hệ thống tự sinh, khách không làm gì cả, nên không có hành động nào để ghi.

**Đây là NULL có cấu trúc, không phải dữ liệu thiếu.** Hai thứ này xử lý ngược nhau:
Vì vậy cần chú ý, nếu hiểu lầm, và xóa đi các cột đó thì có thể làm mất dữ liệu chi phí huy động vốn của ngân hàng

Vì vậy tôi điền nhãn rõ nghĩa như `INTEREST_CREDIT`, không dùng `Unknown`.


Quét đồng nhất: thống kê cột số

In [57]:
for ten, df in bang.items():
    so = df.select_dtypes(include='number')
    if so.empty:
        continue
    print('=' * 90)
    print(ten.upper())
    print('=' * 90)
    display(so.describe().T)

ACCOUNT


,count,mean,std,min,25%,50%,75%,max
account_id,4500.0,2786.067556,2313.811984,1.0,1182.75,2368.0,3552.25,11382.0
district_id,4500.0,37.310444,25.177217,1.0,13.00,38.0,60.00,77.0
date,4500.0,951654.608667,14842.188377,930101.0,931227.00,960102.0,961101.00,971229.0


CARD


,count,mean,std,min,25%,50%,75%,max
card_id,892.0,480.855381,306.933982,1.0,229.75,456.5,684.25,1247.0
disp_id,892.0,3511.862108,2984.373626,9.0,1387.00,2938.5,4459.50,13660.0


CLIENT


,count,mean,std,min,25%,50%,75%,max
client_id,5369.0,3359.011920,2832.911984,1.0,1418.0,2839.0,4257.0,13998.0
birth_number,5369.0,535114.970013,172895.618429,110820.0,406009.0,540829.0,681013.0,875927.0
district_id,5369.0,37.310114,25.043690,1.0,14.0,38.0,60.0,77.0


DISP


,count,mean,std,min,25%,50%,75%,max
disp_id,5369.0,3337.097970,2770.418826,1.0,1418.0,2839.0,4257.0,13690.0
client_id,5369.0,3359.011920,2832.911984,1.0,1418.0,2839.0,4257.0,13998.0
account_id,5369.0,2767.496927,2307.843630,1.0,1178.0,2349.0,3526.0,11382.0


DISTRICT


,count,mean,std,min,25%,50%,75%,max
A1,77.0,39.000000,22.371857,1.00,20.00,39.0,58.00,77.0
A4,77.0,133884.896104,136913.464909,42821.00,85852.00,108871.0,139012.00,1204953.0
A5,77.0,48.623377,32.741829,0.00,22.00,49.0,71.00,151.0
A6,77.0,24.324675,12.780991,0.00,16.00,25.0,32.00,70.0
A7,77.0,6.272727,4.015222,0.00,4.00,6.0,8.00,20.0
A8,77.0,1.727273,1.008338,0.00,1.00,2.0,2.00,5.0
A9,77.0,6.259740,2.435497,1.00,5.00,6.0,8.00,11.0
A10,77.0,63.035065,16.221727,33.90,51.90,59.8,73.50,100.0
A11,77.0,9031.675325,790.202347,8110.00,8512.00,8814.0,9317.00,12541.0
A13,77.0,3.787013,1.908480,0.43,2.31,3.6,4.79,9.4


LOAN


,count,mean,std,min,25%,50%,75%,max
loan_id,682.0,6172.466276,682.579279,4959.0,5577.5,6176.5,6752.5,7308.0
account_id,682.0,5824.162757,3283.512681,2.0,2967.0,5738.5,8686.0,11362.0
date,682.0,963027.910557,14616.406049,930705.0,950704.5,970206.5,971212.5,981208.0
amount,682.0,151410.175953,113372.406310,4980.0,66732.0,116928.0,210654.0,590820.0
duration,682.0,36.492669,17.075219,12.0,24.0,36.0,48.0,60.0
payments,682.0,4190.664223,2215.830344,304.0,2477.0,3934.0,5813.5,9910.0


ORDERS


,count,mean,std,min,25%,50%,75%,max
order_id,6471.0,3.377820e+04,3.737682e+03,29401.0,31187.5,32988.0,34785.5,46338.0
account_id,6471.0,2.962303e+03,2.518503e+03,1.0,1223.0,2433.0,3645.5,11362.0
account_to,6471.0,4.939904e+07,2.888356e+07,399.0,24159184.0,49756062.0,74000447.5,99994199.0
amount,6471.0,3.280636e+03,2.714475e+03,1.0,1241.5,2596.0,4613.5,14882.0


TRANS


,count,mean,std,min,25%,50%,75%,max
trans_id,1056320.0,1.335311e+06,1.227487e+06,1.0,430262.75,858506.5,2.060979e+06,3682987.0
account_id,1056320.0,2.936867e+03,2.477345e+03,1.0,1204.00,2434.0,3.660000e+03,11382.0
date,1056320.0,9.656748e+05,1.394535e+04,930101.0,960116.00,970410.0,9.802280e+05,981231.0
amount,1056320.0,5.924146e+03,9.522735e+03,0.0,135.90,2100.0,6.800000e+03,87400.0
balance,1056320.0,3.851833e+04,2.211787e+04,-41125.7,22402.50,33143.4,4.960362e+04,209637.0
account,295389.0,4.567092e+07,3.066340e+07,0.0,17828584.00,45750951.0,7.201341e+07,99994199.0


Nhận xét

trans.balance ( số dư sau giao dịch) có giá trị âm: cần kiểm tra xem số dư âm trong dữ liệu này có ý nghĩa gì hay là do dữ liệu có vấn đề.
client.birth_number: cột này đã có dấu hiệu bất thường từ ban đầu, nên cần kiểm tra kỹ hơn để hiểu con số này đang được lưu theo dạng nào và có phải ngày sinh hay không.

Số dư âm

In [58]:
# Lấy những giao dịch có số dư âm
so_du_am = trans[trans['balance'] < 0]

# Có bao nhiêu giao dịch có số dư âm?
print('Số giao dịch có số dư âm:', len(so_du_am))

# Chiếm bao nhiêu % tổng số giao dịch?
print('Tỷ lệ:', f'{len(so_du_am) / len(trans) * 100:.2f}%')

# Có bao nhiêu tài khoản từng có số dư âm?
print('Số tài khoản có số dư âm:', so_du_am['account_id'].nunique())

# Số dư âm thấp nhất là bao nhiêu?
print('Số dư thấp nhất:', so_du_am['balance'].min())

Số giao dịch có số dư âm: 2999
Tỷ lệ: 0.28%
Số tài khoản có số dư âm: 288
Số dư thấp nhất: -41125.7


Có 2.999 giao dịch có số dư âm, chiếm khoảng 0,28% tổng số giao dịch. Các giao dịch này thuộc 288 tài khoản, và số dư thấp nhất là -41.125,7.

Kiểm tra xem những giao dịch có số dư âm thường thuộc loại k_symbol nào.

In [59]:
print('k_symbol cua cac dong so du am:')
print(so_du_am['k_symbol'].value_counts(dropna=False).to_string())

k_symbol cua cac dong so du am:
k_symbol
NaN            939
SANKC. UROK    739
UROK           456
SLUZBY         402
SIPO           265
               110
POJISTNE        54
UVER            34


## Nhận xét 

Số dư âm không nhất thiết là lỗi dữ liệu. Nó có thể phản ánh trường hợp tài khoản bị thấu chi, tức là khách hàng đã sử dụng nhiều tiền hơn số dư hiện có.

Vì vậy, đây là một thông tin có ý nghĩa về mặt nghiệp vụ, không nên xem là dữ liệu cần xóa. Những tài khoản thường xuyên có số dư âm có thể là nhóm cần được theo dõi thêm.

Quyết định: giữ nguyên dữ liệu, không lọc bỏ và không đổi số dư âm thành 0.

`birth_number`

Bảng `client` chỉ có 3 cột, không có ngày sinh và không có giới tính. Nhưng `birth_number` trông giống dạng ngày `YYMMDD` đã gặp ở đầu mối 2.

Áp dụng lại đúng hàm kiểm tra đã viết lúc đó:

In [60]:
kiem_tra_yymmdd('client.birth_number', client['birth_number'])

client.birth_number    nam 11-87   thang 1-62   ngay 1-31


Phần tháng vượt quá hơn 12 tháng. 

Phần tháng vượt quá 12 — phép thử `YYMMDD` thất bại ở đây, trong khi nó đã thoả với ba cột `date` trước đó.

Nhưng `min = 1, max = 62` chưa đủ để kết luận gì. 

In [61]:
thang = (client['birth_number'] // 100) % 100

thang.value_counts().sort_index()

birth_number
1     233
2     236
3     218
4     223
5     246
6     217
7     239
8     232
9     209
10    231
11    216
12    224
51    221
52    192
53    209
54    225
55    221
56    195
57    226
58    238
59    260
60    235
61    210
62    213
Name: count, dtype: int64

Phân bố tách thành **hai cụm rời nhau**, không phải dải liên tục. Đo khoảng cách và kích thước hai cụm.

In [62]:
cum_thap = thang[thang <= 12]
cum_cao  = thang[thang > 12]

print(f'Cum thap: {cum_thap.min()}-{cum_thap.max()}   co {len(cum_thap)} nguoi')
print(f'Cum cao : {cum_cao.min()}-{cum_cao.max()}   co {len(cum_cao)} nguoi')
print()
print('Do lech giua hai cum:', cum_cao.min() - cum_thap.min())
print(f'Ty le: {100 * len(cum_thap) / len(client):.1f}% / {100 * len(cum_cao) / len(client):.1f}%')

Cum thap: 1-12   co 2724 nguoi
Cum cao : 51-62   co 2645 nguoi

Do lech giua hai cum: 50
Ty le: 50.7% / 49.3%


## Nhận xét 


1. Hai cụm cách nhau **đúng 50** — một số tròn, không phải nhiễu
2. Cụm cao trải **đúng 12 giá trị**, giống hệt cụm thấp — vẫn là 12 tháng
3. Hai cụm **gần bằng nhau về số lượng**, xấp xỉ 50/50


Tra cứu với từ khoá *"Czech birth number encoding"* xác nhận: đây là mã định danh cá nhân của Séc (*rodné číslo*), quy ước cộng 50 vào phần tháng đối với nữ. Một con số mang được cả ngày sinh lẫn giới tính.

**Lưu ý:** dữ liệu chứng minh được *"một trong hai nhóm bị cộng 50"*, nhưng **không** chứng minh được nhóm đó là nữ hay nam. 

**Quyết định: tách `birth_number` thành hai cột** — `birth_date` và `gender`.



# TỔNG KẾT
## Bảng quyết định làm sạch

| # | Vấn đề | Quyết định |
|---|---|---|
| 1 | `district.A12`, `A15` chứa `?` | Để NULL, không điền thay thế |
| 2 | Cột ngày là số nguyên `YYMMDD` | Đổi sang kiểu `DATE` |
| 3 | `disp` có 2 dòng cho 1 tài khoản | Lọc `type = 'OWNER'` |
| 4 | `operation` rỗng 183.114 dòng | Điền nhãn `INTEREST_CREDIT` |
| 5 | `birth_number` mã hoá giới tính | Tách thành `birth_date` + `gender` |
| 6 | `trans.type` có 3 giá trị | Gộp `VYBER` vào `VYDAJ`, còn 2 |
| 7 | `k_symbol` có 2 kiểu rỗng | `NULLIF(TRIM(k_symbol), '')` |
| 8 | Số dư âm | **Giữ nguyên**  |
